File: nodeServer/bluetoothNodeServer.py
#Python version
import os
import subprocess
import time
import datetime
import rethinkdb as rethinkdb
from rethinkdb.errors import RqlRuntimeError, RqlDriverError

databaseConnection = rethinkdb.connect(host='127.0.0.2', port=28015)
devices = rethinkdb.db('bluetooth_node').table('bluetooth_devices').run(databaseConnection)
#run indefinitely
while (True):
    #pull fresh list of devices
    changeFeed = rethinkdb.db('bluetooth_node').table('bluetooth_devices').changes().run(databaseConnection)
    #check for changes in the changefeed
    changesFound = False
    for change in changeFeed:
        changesFound = True
        break
    #if there are changes, do a fresh pull of devices
    if (changesFound):
        devices = rethinkdb.db('bluetooth_node').table('bluetooth_devices').run(databaseConnection)
    for device in devices:
        #poll for device on node server
        output = subprocess.check_output(['hcitool', 'name', device['device_address']])
        #if present, write timestamp to database
        if (output != ''):
            phoneName = output.split('\n')[0]
            exactTime = time.time()
            timeStamp = datetime.datetime.fromtimestamp(exactTime).strftime('%Y-%m-%d %H:%M:%S')
            rethinkdb.db('bluetooth_node').table('bluetooth_connections').insert({'bluetooth_address' : device['device_address'], 'phone_name' : phoneName, 'time_stamp' : timeStamp}).run(databaseConnection)
    time.sleep(1)

File: node/bluetoothNodeServer.js
//Node.js version
var rethinkdb = require('rethinkdb');

//open connection to database
//callback pulls a list of devices
var databaseConnection = null;
rethinkdb.connect({host: '127.0.0.2', port: 28015}, function(err, conn) {
    if (err) throw err;
    databaseConnection = conn;
    rethinkdb.db('bluetooth_node').table('bluetooth_devices').run(databaseConnection, function(err, cursor) {
        if (err) throw err;
        cursor.toArray(function(err, result) {
            if(err) throw err;
            var devices = result;
            //run indefinitely
            while (true) {
                //check changefeed
                rethinkdb.db('bluetooth_node').table('bluetooth_devices').changes().run(databaseConnection, function(err, cursor) {
                    console.log('here');
                    changeFeed = cursor;
                    changesFound = false;
                    for (var change in changeFeed) {
                        changesFound = true;
                    }
                    if (changesFound) {
                        rethinkdb.db('bluetooth_node').table('bluetooth_devices').run(databaseConnection, function(err, cursor) {
                            if (err) throw err;
                            cursor.toArray(function(err, result) {
                                if(err) throw err;
                                var devices = result;
                            });
                        });
                    }
                    for (var device in devices) {

                    }
                });
            }
        });
    });
});